### PR TITLE
[NFCI][SYCL] Refactor device selection in platform_impl.cpp

### PR DESCRIPTION
**Pull request by aelovikov-intel**
Mostly "early continue" and use the same idioms for similar things.

### CODE REVIEW

File: sycl/source/detail/platform_impl.cpp [hunk 1]
```diff
@@ -392,116 +366,101 @@ static std::vector<device> amendDeviceAndSubDevices(
...
+        continue;
+      }
+
+      if (!supportsSubPartitioning) {
+        if (target.DeviceNum ||
+            (target.DeviceType &&
+             (target.DeviceType.value() != info::device_type::all))) {
+          // This device was specifically requested and yet is not
+          // partitionable.
+          std::cout << "device is not partitionable: " << target << std::endl;
```
Comment (maarquitos14) [inline sycl/source/detail/platform_impl.cpp:408]: I understand these and other prints are actually diagnosing errors (e.g. trying to partition a device that is not partitionable). Is my understanding correct? If so, I wonder if just printing and keep going as if nothing happened is good enough as a diagnostic.

Comment (aelovikov-intel) [reply sycl/source/detail/platform_impl.cpp:408]: They were here before me, any changes in the behavior should go into a separate PR :)

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/source/detail/platform_impl.cpp (hunk 1 above)
Comment (aelovikov-intel) [reply sycl/source/detail/platform_impl.cpp:408]: I honestly would expect an exception thrown, but maybe @cperkinsintel had his reasons to implement it this way (I assume it was added as part of his major work in this area).

**Pull request merged**